src/utils/offlineRequestHandler.ts

```diff
@@ -0,0 +1,392 @@
+        this.replayRequestsInParallel(submitStepRequests)
+      ]);
+
+      const registerEventRequests = allRequests.filter(
+        (req) => req.type === 'registerEvent'
+      );
+      const stepKeys = Array.from(
+        new Set(registerEventRequests.map((req) => req.stepKey))
+      );
+
+      for (const stepKey of stepKeys) {
+        const requestsForStep = registerEventRequests.filter(
+          (req) => req.stepKey === stepKey
+        );
+        requestsForStep.sort((a, b) => a.timestamp - b.timestamp);
+        await this.replayRequestsInParallel(requestsForStep);
+      }
+
+      this.onlineSignals.forEach((signal) => signal());
+      this.onlineSignals = [];
+    } finally {
+      this.isReplayingRequests = false;
+
+      // Check if there are any new requests in IndexedDB and trigger replayRequests again
+      if (navigator.onLine) {
+        const { store } = await this.getDbTransaction('readonly');
+        const count = await new Promise<number>((resolve) => {
+          const request = store.count();
+          request.onsuccess = () => resolve(request.result);
+        });
+
+        if (count > 0) {
+          await this.replayRequests();
+        }
+      }
+    }
+  }
+
+  private async replayRequestsInParallel(requests: SerializedRequest[]) {
+    await Promise.all(
+      requests.map(async (request) => {
+        const { url, method, headers, body, bodyType } = request;
+        const reconstructedBody = this.reconstructBody(body, bodyType);
+        const fetchOptions: RequestInit = {
+          method,
+          headers: JSON.parse(headers),
+          body: reconstructedBody,
+          cache: 'no-store',
+          keepalive: ['POST', 'PATCH', 'PUT'].includes(method)
+        };
+
+        let attempts = 0;
+        let success = false;
+
+        while (!success && attempts < this.maxRetryAttempts) {
+          try {
+            const response = await fetch(url, fetchOptions);
+            await this.checkResponseSuccess(response);
+            success = true;
+            await this.removeRequest();
+          } catch (error: any) {
+            if (
+              (this.ignoreNetworkErrors?.current ||
+                TYPE_MESSAGES_TO_IGNORE.includes(error.message)) &&
+              error instanceof TypeError
+            ) {
+              return;
+            }
+            attempts++;
+            await this.delay(this.retryDelayMs);
+          }
+
+          if (!navigator.onLine) {
+            // If offline, block until requests are replayed from online event handler
+            await new Promise((resolve) => this._addOnlineSignal(resolve));
+            return;
+          }
+
+          if (attempts === this.maxRetryAttempts) {
+            break;
+          }
+        }
+      })
+    );
+  }
+
+  private async removeRequest() {
+    const { store } = await this.getDbTransaction('readwrite');
+    await store.clear();
+  }
+
+  private async checkResponseSuccess(response: any) {
```

Comment:
Can you go through the PR again and re-review any duplicate code that we can unify? e.g. this is the same code path as in the client, and they can probably be unified